akka-actor/src/main/scala/akka/io/TcpConnection.scala

```diff
@@ -389,6 +398,12 @@ private[io] abstract class TcpConnection(val tcp: TcpExt, val channel: SocketCha
           if (data eq remainingData) this
           else new PendingBufferWrite(commander, data, ack, buffer, tail) // copy with updated remainingData
 
+        } else if (data.nonEmpty && buffer.isReadOnly) {
+          val head = data.headByteString
+          val remaining = data.drop(head.length)
```

Comment:
there could be because of the sparse headByteString logic.

normally the chain of 'PendingBufferWrite's would just acquire a temp rw-bytebuffer when its created, and pass that buffer to each subsequent PendingBufferWrite that could be made (or just hammers on doWrite inline until its done), the calls release on the buffer after everything.

This change breaks that assumption and on each instance of PendingBufferWrite you're not sure if you have a rw or ro bytebuffer, and thats why the factory method is called again as it checks if the next 'bytestring' chunk (any ByteString1/1C thats at the head of a ByteStrings) can be wrapped as a ro-buffer. 

The way this logic ends up panning out is that it will use read-only buffers as long as possible, and may switch to a rw buffer (but never back to a ro-bytebuffer). The thinking around this was that i didnt want to modify with the logic around releasing buffers from PendingBufferWrite. The connection is set up to only release the last buffer in the 'chain' of PendingBufferWrite, so I wanted to make sure that if one was pulled from the pool, it stayed there for the whole chain.

I'm going to fix this. It is fragile, and the fact I had to write this much about it probably means its super murky. Plus this logic as it is can enable really tiny writes which isn't ideal when making kernel calls.